filebeat/input/file.go

```diff
@@ -72,7 +77,37 @@ func mergeJSONFields(f *FileEvent, event common.MapStr) {
 	if f.JSONConfig.KeysUnderRoot {
 		for k, v := range f.JSONFields {
 			if f.JSONConfig.OverwriteKeys {
-				event[k] = v
+				if k == "@timestamp" {
+					vstr, ok := v.(string)
+					if !ok {
+						logp.Err("JSON: Won't overwrite @timestamp because value is not string")
+						event[jsonErrorKey] = "@timestamp not overwritten (not string)"
+						continue
+					}
+					// @timestamp must be of time common.Time
+					ts, err := common.ParseTime(vstr)
+					if err != nil {
+						logp.Err("JSON: Won't overwrite @timestamp because of parsing error: %v", err)
+						event[jsonErrorKey] = fmt.Sprintf("@timestamp not overwritten (parse error on %s)", vstr)
+						continue
+					}
+					event[k] = ts
+				} else if k == "type" {
+					vstr, ok := v.(string)
+					if !ok {
+						logp.Err("JSON: Won't overwrite type because value is not string")
+						event[jsonErrorKey] = "type not overwritten (not string)"
+						continue
+					}
+					if len(vstr) == 0 || vstr[0] == '_' {
+						logp.Err("JSON: Won't overwrite type because value is empty or starts with an underscore")
+						event[jsonErrorKey] = fmt.Sprintf("type not overwritten (invalid value [%s])", vstr)
+						continue
+					}
+					event[k] = vstr
+				} else {
+					event[k] = v
```

Comment:
Then it gets overwritten. IMHO that's fine, because `keys_under_root` is not the default and we usually recommend people not to use it. When people use it, it's implied that "I'm aware that everything can get overwritten".